node/full_node_integration_test.go

```diff
@@ -353,6 +353,130 @@ func testSingleAggreatorSingleFullNodeSingleLightNode(t *testing.T) {
 	assert.Equal(n1h, n3h, "heights must match")
 }
 
+func checkAppCalls(assert *assert.Assertions, app *mocks.Application, expected []int) {
+	generateFraudProofCnt := 0
+	verifyFraudProofCnt := 0
+	for _, call := range app.Calls {
+		switch call.Method {
+		case "GenerateFraudProof":
+			generateFraudProofCnt++
+		case "VerifyFraudProof":
+			verifyFraudProofCnt++
+		}
+	}
+	assert.Equal(expected[0], generateFraudProofCnt)
+	assert.Equal(expected[1], verifyFraudProofCnt)
+}
+
+func testSingleAggreatorSingleFullNodeFraudProofGossip(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	var wg sync.WaitGroup
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	clientNodes := 1
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+
+	for _, app := range apps {
+		app.On("VerifyFraudProof", mock.Anything).Return(abci.ResponseVerifyFraudProof{Success: true}).Run(func(args mock.Arguments) {
+			wg.Done()
+		}).Once()
+	}
+
+	aggNode := nodes[0]
+	fullNode := nodes[1]
+
+	wg.Add(clientNodes + 1)
+	require.NoError(aggNode.Start())
+	time.Sleep(2 * time.Second)
+	require.NoError(fullNode.Start())
+
+	wg.Wait()
+	// aggregator should have 0 GenerateFraudProof calls and 1 VerifyFraudProof calls
+	checkAppCalls(assert, apps[0], []int{0, 1})
+	// fullnode should have 1 GenerateFraudProof calls and 1 VerifyFraudProof calls
+	checkAppCalls(assert, apps[1], []int{1, 1})
+
+	n1Frauds, err := aggNode.fraudService.Get(aggCtx, types.StateFraudProofType)
+	require.NoError(err)
+	aggCancel()
+	require.NoError(aggNode.Stop())
+
+	n2Frauds, err := fullNode.fraudService.Get(aggCtx, types.StateFraudProofType)
+	require.NoError(err)
+	cancel()
+	require.NoError(fullNode.Stop())
+
+	assert.Equal(len(n1Frauds), 1, "number of fraud proofs received via gossip should be 1")
+	assert.Equal(len(n2Frauds), 1, "number of fraud proofs received via gossip should be 1")
+	assert.Equal(n1Frauds, n2Frauds, "the received fraud proofs after gossip must match")
+}
+
+func testSingleAggreatorTwoFullNodeFraudProofSync(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	var wg sync.WaitGroup
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	clientNodes := 2
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+
+	for _, app := range apps {
+		app.On("VerifyFraudProof", mock.Anything).Return(abci.ResponseVerifyFraudProof{Success: true}).Run(func(args mock.Arguments) {
+			wg.Done()
+		}).Once()
+	}
+
+	aggNode := nodes[0]
+	fullNode1 := nodes[1]
+	fullNode2 := nodes[2]
+
+	wg.Add(clientNodes)
+	require.NoError(aggNode.Start())
+	time.Sleep(2 * time.Second)
```

Comment:
can you add a comment explaining this? what is the error or non-determinism that happens if we don't sleep between calls?